File: 18_哈希/ZobristHashing/AllCountKChecker/AllCountMultipleOfKChecker.py
from collections import defaultdict
from random import randint
from typing import List


def countSubarrayWithFrequencyMultipleOfK(arr: List[int], k: int) -> int:
    """统计满足`每个元素出现的次数均为k的倍数`条件的子数组的个数."""
    n = len(arr)
    if n == 0 or k <= 0 or k > n:
        return 0

    pool = defaultdict(lambda: randint(1, (1 << 61) - 1))
    id_ = defaultdict(lambda: len(id_))
    arr = [id_[v] for v in arr]
    counter = [0] * len(id_)
    random = [pool[v] for v in arr]
    preSum = [0] * (n + 1)
    for i, v in enumerate(arr):
        preSum[i + 1] = preSum[i]
        preSum[i + 1] -= counter[v] * random[i]
        counter[v] = (counter[v] + 1) % k
        preSum[i + 1] += counter[v] * random[i]


class AllCountMultipleOfKChecker:
    """
    判断数据结构中每个数出现的次数是否均k的倍数.
    如果为空集合,则返回True.
    """

    _poolSingleton = defaultdict(lambda: randint(1, (1 << 61) - 1))

    __slots__ = ("_hash", "_modCounter", "_k")

    def __init__(self, k: int) -> None:
        self._hash = 0
        self._modCounter = defaultdict(int)
        self._k = k

    def add(self, x: int) -> None:
        count, random = self._modCounter[x], self._poolSingleton[x]
        self._hash -= count * random
        count += 1
        if count == self._k:
            count = 0
        self._hash += count * random
        self._modCounter[x] = count

    def remove(self, x: int) -> None:
        """删除前需要保证x在集合中存在."""
        count, random = self._modCounter[x], self._poolSingleton[x]
        self._hash -= count * random
        count -= 1
        if count == -1:
            count = self._k - 1
        self._hash += count * random
        self._modCounter[x] = count

    def query(self) -> bool:
        return self._hash == 0

    def getHash(self) -> int:
        return self._hash

    def clear(self) -> None:
        self._hash = 0
        self._modCounter.clear()

File: 18_哈希/ZobristHashing/AllCountKChecker/AllCountMultipleOfKChecker.go
package main

import (
	"fmt"
	"math/rand"
)

func main() {
	C := NewAllCountMultipleOfKChecker(3)
	fmt.Println(C.Query())
	C.Add(1)
	fmt.Println(C.Query())
	C.Add(1)
	fmt.Println(C.Query())
	C.Add(1)
	fmt.Println(C.Query())
	C.Add(2)
	fmt.Println(C.Query())
	C.Remove(2)
	fmt.Println(C.Query())
}

type T int

// 判断数据结构中每个数出现的次数是否均k的`倍数`.
type AllCountMultipleOfKChecker struct {
	pool    map[_pair]uint64
	hash    uint64
	k       int
	counter map[T]int
}

func NewAllCountMultipleOfKChecker(k int) *AllCountMultipleOfKChecker {
	res := &AllCountMultipleOfKChecker{}
	res.pool = make(map[_pair]uint64)
	res.k = k
	res.counter = make(map[T]int)
	return res
}

func (c *AllCountMultipleOfKChecker) Add(v T) {
	count := c.counter[v]
	c.hash ^= c._hash(v, count)
	count++
	if count == c.k {
		count = 0
	}
	c.counter[v] = count
	c.hash ^= c._hash(v, count)
}

// 删除前需要保证v在集合中存在.
func (c *AllCountMultipleOfKChecker) Remove(v T) {
	count := c.counter[v]
	c.hash ^= c._hash(v, count)
	count--
	if count == -1 {
		count = c.k - 1
	}
	c.counter[v] = count
	c.hash ^= c._hash(v, count)
}

// 询问数据结构中每个数出现的次数是否均k的倍数.
func (c *AllCountMultipleOfKChecker) Query() bool {
	return c.hash == 0
}

func (c *AllCountMultipleOfKChecker) GetHash() uint64 {
	return c.hash
}

func (c *AllCountMultipleOfKChecker) _hash(v T, countMod int) uint64 {
	p := _pair{v, countMod}
	if res, ok := c.pool[p]; ok {
		return res
	}
	res := c._randUint61()
	c.pool[p] = res
	return res
}

// [1,1<<61-1]
func (c *AllCountMultipleOfKChecker) _randUint61() uint64 {
	return rand.Uint64()%(1<<61-1) + 1
}

type _pair struct {
	value    T
	countMod int
}
